src/encrypt.ts

```diff
@@ -0,0 +1,152 @@
+// external modules
+import { createCipheriv } from 'crypto-browserify';
+
+// local
+import { Cipher } from 'crypto';
+import toBuffer from 'typedarray-to-buffer';
+import {
+  PenumbraDecryptionInfo,
+  PenumbraEncryptedFile,
+  PenumbraEncryptionOptions,
+  PenumbraFile,
+} from './types';
+
+// utils
+import { emitProgress, toBuff } from './utils';
+
+/* tslint:disable completed-docs */
+
+// external modules
+
+/**
+ * Encrypts a readable stream
+ *
+ * @param rs - A readable stream of encrypted data
+ * @param cipher - The crypto module's cipher
+ * @param contentLength - The content length of the file, in bytes
+ * @param url - The URL to read the encrypted file from (only used for the event emitter)
+ * @returns A readable stream of decrypted data
+ */
+export function encryptStream(
+  rs: ReadableStream,
+  cipher: Cipher,
+  contentLength: number,
+): ReadableStream {
+  let totalBytesRead = 0;
+
+  // TransformStreams are supported
+  if ('TransformStream' in self) {
+    return rs.pipeThrough(
+      // eslint-disable-next-line no-undef
+      new TransformStream({
+        transform: async (chunk, controller) => {
+          const bufferChunk = toBuffer(chunk);
+
+          // Encrypt chunk and send it out
+          const encryptedChunk = cipher.update(bufferChunk);
+          controller.enqueue(encryptedChunk);
+
+          // Emit a progress update
+          totalBytesRead += bufferChunk.length;
+          emitProgress(
+            'encrypt',
+            totalBytesRead,
+            contentLength,
+            '[encrypted file]',
+          );
+        },
+      }),
+    );
+  }
+
+  // TransformStream not supported, revert to ReadableStream
+  const reader = rs.getReader();
+  return new ReadableStream({
+    /**
+     * Controller
+     */
+    start(controller) {
+      /**
+       * Push on
+       */
+      function push(): void {
+        reader.read().then(({ done, value }) => {
+          if (done) {
+            controller.close();
+            return;
+          }
+
+          const chunk = toBuffer(value);
+
+          // Encrypt chunk
+          const encryptedChunk = cipher.update(chunk);
+
+          // Emit a progress update
+          totalBytesRead += chunk.length;
+          emitProgress(
+            'encrypt',
+            totalBytesRead,
+            contentLength,
+            '[encrypted file]',
+          );
+
+          controller.enqueue(encryptedChunk);
+          push();
+        });
+      }
+      push();
+    },
+  });
+}
+
+/** Encrypt a buffer */
+export function encryptBuffer(
+  buffer: ArrayBuffer,
+  cipher: Cipher,
+): ArrayBuffer {
+  console.error('penumbra encryptBuffer() is not yet implemented');
+  return new ArrayBuffer(10);
+}
+
+// Minimum IV randomness set by NIST.
+// Should this be 16 to align with 256-bit byte boundaries?
```

Comment:
96 bit (12 byte) IVs are [recommended](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), primarily for interoperability and efficiency. The GCM implementation will slice it down to 12 bits anyway, so choosing a larger IV will not increase entropy. If the IV were a counter, then increasing length could improve uniqueness (which is all that's important for IVs in GCM mode), but it's not the case here

https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf

Reply:
To elaborate... GCM fails when the same IV and key are used multiple times. So for each key used, it's important that the same IV isn't reused. The IV here is like the birthday problem with 2^96 birthdays so it's implausible to repeat with a good CSPRNG... especially given keys are not thoroughly reused.